skbio/stats/composition.py

```diff
@@ -0,0 +1,410 @@
+
+    """
+    mat = np.atleast_2d(mat)
+    if mat.ndim > 2:
+        raise ValueError("Input matrix can only have two dimensions or less")
+    if np.any(mat < 0):
+        raise ValueError("Cannot have negative proportions")
+    if not np.all(np.isclose(mat.sum(axis=1), 1)):
+        raise ValueError("Rows need to sum up to 1")
+
+    z_mat = (mat == 0)
+
+    num_samps, num_feats = mat.shape
+    tot = z_mat.sum(axis=1, keepdims=True)
+
+    if delta is None:
+        delta = (1. / num_feats)**2
+
+    zcnts = 1 - tot * delta
+    mat = np.where(z_mat, delta, zcnts * mat)
+    return mat.squeeze()
+
+
+def perturb(x, y):
+    r"""
+    Performs the perturbation operation
+
+    This operation is defined as
+    :math:`x \oplus y = C[x_1 y_1, ..., x_D y_D]`
+
+    :math:`C[x]` is the closure operation defined as
+    :math:`C[x] = [\frac{x_1}{\sum x},...,\frac{x_D}{\sum x}]`
+    for some :math:`D` dimensional real vector :math:`x` and
+    :math:`D` is the number of components for every composition.
+
+    Parameters
+    ----------
+    x : array_like, float
+        a matrix of proportions where
+        rows = compositions and
+        columns = components
+    y : array_like, float
+        a matrix of proportions where
+        rows = compositions and
+        columns = components
+
+    Returns
+    -------
+    numpy.ndarray, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    Notes
+    -----
+    - All of the values in x and y must be greater than zero
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skbio.stats.composition import perturb
+    >>> x = np.array([.1,.3,.4, .2])
+    >>> y = np.array([1./6,1./6,1./3,1./3])
+    >>> perturb(x,y)
+    array([ 0.0625,  0.1875,  0.5   ,  0.25  ])
+
+    """
+    x = np.asarray(x, dtype=np.float64)
+    y = np.asarray(y, dtype=np.float64)
+    if np.any(x < 0) or np.any(y < 0):
+        raise ValueError("Cannot have negative proportions")
+    return closure(x * y)
+
+
+def perturb_inv(x, y):
+    r"""
+    Performs the inverse perturbation operation
+
+    This operation is defined as
+    :math:`x \ominus y = C[x_1 y_1^{-1}, ..., x_D y_D^{-1}]`
+
+    :math:`C[x]` is the closure operation defined as
+    :math:`C[x] = [\frac{x_1}{\sum x},...,\frac{x_D}{\sum x}]`
+    for some :math:`D` dimensional real vector :math:`x` and
+    :math:`D` is the number of components for every composition.
+
+    Parameters
+    ----------
+    x : array_like
+        a matrix of proportions where
+        rows = compositions and
+        columns = components
+    y : array_like
+        rows = compositions and
+        columns = components
+
+    Returns
+    -------
+    numpy.ndarray, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    Notes
+    -----
+    - All of the values in x and y must be greater than zero
```

Comment:
Either the docstring or the checks are not correct regarding dealing with zeros, here and in other functions.

In fact, validation is the same for all functions, so it should be extracted to a helper function. Either improving `closure` and calling that for each input array or having `_validate` that just converts to arrays and checks signs and raises the appropriate error.

I like the 1st option because it seems that this module would error out in a more user friendly way if someone tries to use this with non-compositional data by accident.